# Apriori

## Importing the libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Data Preprocessing

In [2]:
dataset = pd.read_csv('Market_Basket_Optimisation.csv', header = None)
dataset.head(3)

0          1        2               3             4   \
0   shrimp    almonds  avocado  vegetables mix  green grapes   
1  burgers  meatballs     eggs             NaN           NaN   
2  chutney        NaN      NaN             NaN           NaN   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN

In [6]:
dataset.shape

(7501, 20)

In [3]:
# Input has to be a list of all items , so preparing data accordingly

In [7]:
transactions =[]

for i in dataset.index:
    transactions.append( dataset.loc[i,:].dropna().to_list() )
    
for i in transactions[:3]:
    print(i)
    print()

['shrimp', 'almonds', 'avocado', 'vegetables mix', 'green grapes', 'whole weat flour', 'yams', 'cottage cheese', 'energy drink', 'tomato juice', 'low fat yogurt', 'green tea', 'honey', 'salad', 'mineral water', 'salmon', 'antioxydant juice', 'frozen smoothie', 'spinach', 'olive oil']

['burgers', 'meatballs', 'eggs']

['chutney']



## Training the Apriori model on the dataset

In [8]:
# WARNING: Make sure to upload the apyori.py file into this Colab notebook before running this cell
from apyori import apriori
rules = apriori(transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2)

## Visualising the results

In [9]:
results = list(rules)

In [10]:
print(results)

[RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]), RelationRecord(items=frozenset({'mushroom cream sauce', 'escalope'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]), RelationRecord(items=frozenset({'pasta', 'escalope'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]), RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0.24

# The data is really Clumsy let's clean it up

In [11]:
results[0]

RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)])

- RelationRecord(
  -  items=frozenset({'chicken', 'light cream'}),support=0.004532728969470737,   >>> Items that are brought together
   - ordered_statistics=[
    -    OrderedStatistic( 
     -   items_base=frozenset({'light cream'}),                                  >>>First the user bought light cream
      -  items_add=frozenset({'chicken'}),                                       >>>User who bought light_cream also bought chicken
       - confidence=0.29059829059829057, lift=4.84395061728395)
        -]
    - )

In [27]:
list(results[0][2][0][0])[0]

'light cream'

In [39]:
results[0][2][0][1]

frozenset({'chicken'})

In [59]:
items=[]
support=[]
antecedent_item = []
consequent_item = []
confidence = []
lift = []
ind = []

for j in range(0,len(results)):
    
    for i in results[j][2]:
        items.append( sorted(list( results[0][0] ) ) )
        support.append( results[j][1] )
        antecedent_item.append(sorted(list(i[0]) ))
        consequent_item.append(sorted(list(i[1] ) )[0] )
        confidence.append(i[2])
        lift.append(i[3])
        ind.append(j)

In [60]:
data = pd.DataFrame({'Items' : items , 'antecedent' : antecedent_item ,'consequent' : consequent_item ,
                    'Support' : support , 'confidence' : confidence , 'lift' : lift , 'Index' : ind})

In [61]:
data.head()

Items              antecedent   consequent   Support  \
0  [chicken, light cream]           [light cream]      chicken  0.004533   
1  [chicken, light cream]  [mushroom cream sauce]     escalope  0.005733   
2  [chicken, light cream]                 [pasta]     escalope  0.005866   
3  [chicken, light cream]         [fromage blanc]        honey  0.003333   
4  [chicken, light cream]         [herb & pepper]  ground beef  0.015998   

   confidence      lift  Index  
0    0.290598  4.843951      0  
1    0.300699  3.790833      1  
2    0.372881  4.700812      2  
3    0.245098  5.164271      3  
4    0.323450  3.291994      4

In [51]:
data.shape

(113, 7)

In [64]:
data[data['consequent'] == 'ground beef' ]

Items                                     antecedent  \
4    [chicken, light cream]                                [herb & pepper]   
5    [chicken, light cream]                                 [tomato sauce]   
16   [chicken, light cream]                           [cereals, spaghetti]   
24   [chicken, light cream]                     [chocolate, herb & pepper]   
27   [chicken, light cream]                       [cooking oil, spaghetti]   
29   [chicken, light cream]                          [eggs, herb & pepper]   
32   [chicken, light cream]                  [french fries, herb & pepper]   
34   [chicken, light cream]                 [frozen vegetables, spaghetti]   
47   [chicken, light cream]                     [grated cheese, spaghetti]   
49   [chicken, light cream]                          [herb & pepper, milk]   
50   [chicken, light cream]                 [herb & pepper, mineral water]   
51   [chicken, light cream]                     [herb & pepper, spaghetti]   
54   [chicken, light cream]                            [pepper, spaghetti]   
57   [chicken, light cream]                      [spaghetti, tomato sauce]   
72   [chicken, light cream]               [chocolate, eggs, mineral water]   
73   [chicken, light cream]  [chocolate, frozen vegetables, mineral water]   
76   [chicken, light cream]      [chocolate, frozen vegetables, spaghetti]   
89   [chicken, light cream]       [frozen vegetables, milk, mineral water]   
92   [chicken, light cream]           [frozen vegetables, milk, spaghetti]   
94   [chicken, light cream]  [frozen vegetables, mineral water, spaghetti]   
108  [chicken, light cream]          [mineral water, olive oil, spaghetti]   
110  [chicken, light cream]           [mineral water, spaghetti, tomatoes]   

      consequent   Support  confidence      lift  Index  
4    ground beef  0.015998    0.323450  3.291994      4  
5    ground beef  0.005333    0.377358  3.840659      5  
16   ground beef  0.003066    0.460000  4.681764     14  
24   ground beef  0.003999    0.441176  4.490183     19  
27   ground beef  0.004799    0.302521  3.078982     21  
29   ground beef  0.004133    0.329787  3.356491     22  
32   ground beef  0.003200    0.461538  4.697422     24  
34   ground beef  0.008666    0.311005  3.165328     26  
47   ground beef  0.005333    0.322581  3.283144     35  
49   ground beef  0.003600    0.391304  3.982597     37  
50   ground beef  0.006666    0.390625  3.975683     38  
51   ground beef  0.006399    0.393443  4.004360     39  
54   ground beef  0.003333    0.337838  3.438428     42  
57   ground beef  0.003066    0.489362  4.980600     44  
72   ground beef  0.003999    0.297030  3.023093     55  
73   ground beef  0.003333    0.342466  3.485530     56  
76   ground beef  0.003066    0.389831  3.967597     57  
89   ground beef  0.003733    0.337349  3.433457     65  
92   ground beef  0.003066    0.370968  3.775616     66  
94   ground beef  0.004399    0.366667  3.731841     67  
108  ground beef  0.003066    0.298701  3.040106     73  
110  ground beef  0.003066    0.328571  3.344117     74

In [65]:
# We all know that pepper,herb,spaghetti go well with ground beef

# The above data also says the same